### PR TITLE
Introduce TokenStore protocol

### PR DESCRIPTION
**Pull request by shamsimam**
This change will have a follow-up PR that introduces pluggability from the config using our Custom Extensions support.

### CODE REVIEW

File: waiter/src/waiter/core.clj [hunk 1]
```diff
@@ -830,7 +833,7 @@
                                 (async/tap state-chan-mult state-chan)
                                 (state/start-service-chan-maintainer
                                   {} instance-rpc-chan state-chan query-app-maintainer-chan start-service remove-service retrieve-channel)))
-   :state-query-chans (pc/fnk [[:state query-app-maintainer-chan scheduler]
+   :state-query-chans (pc/fnk [[:state query-app-maintainer-chan]
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj]: Minor cleanup of unused variable.

File: waiter/src/waiter/core.clj [hunk 2]
```diff
@@ -163,7 +163,6 @@
                   (cid/ensure-correlation-id nested-response get-request-cid)
                   nested-response)))))))))
 
-
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj]: Removed extra line (unrelated to this PR)

File: waiter/src/waiter/token.clj [hunk 3]
```diff
@@ -157,6 +169,25 @@
                   token-set (->> tokens (map :token) set)]
               (kv/store kv-store owner-key token-set))))))))
 
+;; A token store which is backed by a key-value store.
+(defrecord KeyValueTokenStore [clock synchronize-fn kv-store]
+
+  TokenStore
+
+  (fetch [_ token include-deleted?]
+    (sd/token->token-description kv-store token :include-deleted include-deleted?))
```
Comment (sradack) [inline waiter/src/waiter/token.clj]: Inline `sd/token->token-description`

File: waiter/src/waiter/token.clj [hunk 4]
```diff
@@ -157,6 +169,25 @@
...
+
+  (store [_ token service-description-template token-metadata]
+    ;; insert the last-update-time if it doesn't exist
+    (let [token-metadata' (merge {"last-update-time" (.getMillis ^DateTime (clock))} token-metadata)]
+      (store-service-description-for-token synchronize-fn kv-store token service-description-template token-metadata')))
```
Comment (sradack) [inline waiter/src/waiter/token.clj]: Inline `store-service-description-for-token`

File: waiter/src/waiter/token.clj [hunk 5]
```diff
@@ -157,6 +169,25 @@
...
+
+  (delete [this token hard-delete?]
+    (let [{:keys [token-metadata]} (fetch this token hard-delete?)]
+      (if-let [owner (get token-metadata "owner")]
+        (delete-service-description-for-token clock synchronize-fn kv-store token owner :hard-delete hard-delete?)
```
Comment (sradack) [inline waiter/src/waiter/token.clj]: Inline `delete-service-description-for-token`

File: waiter/src/waiter/token.clj (hunk 4 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: Inlining requires moving the procol inside the `let` block that defines `token-lock`. Protocols don't appear to recognize variables defined in let blocks, I get the following error:
`Unable to resolve symbol: token-lock in this context, compiling:(waiter/token.clj:138:9)`.

I wrote this simple example and had the same error:
```
(defprotocol P (foo [this]))
(let [x 1]
  (defrecord R [] P (foo [_] (println x))))
=> 
CompilerException java.lang.RuntimeException: Unable to resolve symbol: x in this context, compiling:(/private/var/folders/wj/c9w0p0r93vl4rl815rspwxqw0000gn/T/form-init7112261126071510316.clj:3:30) 
```

File: waiter/src/waiter/token.clj (hunk 5 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: Same `token-lock` issue.

File: waiter/src/waiter/token.clj (hunk 3 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: `sd/token->token-description` and `sd/token->service-description-template` are closely related.

**Pull request closed without merging**